File: exercises/python/03_class.py

import pandas as pd


# Cargar datos

ventas = pd.read_csv(
    "datasets/superstore.csv",
    encoding="latin-1"
)

# ventas.columns = ventas.columns.str.replace(" ","_")

ventas_categoria = (
    ventas
    .groupby("Category")["Sales"]
    .sum()
    .sort_values(ascending=False)
)
print(ventas_categoria)

conteo_ordenes = ventas["Category"].value_counts()

print("\n Conteo de transacciones por categoría:")
print(conteo_ordenes)


ganancia_region = (
    ventas
    .groupby("Region")["Profit"]
    .sum()
    .sort_values(ascending=False)
)


print("\nGanancia por región")
print(ganancia_region)

cuadro_categoria = (
    ventas
    .groupby("Category")
    .agg(
        Ventas_Totales=("Sales", "sum"),
        Ganancia_Total=("Profit", "sum"),
        Unidades_Vendidas=("Quantity", "sum"),
        Ticket_Promedio=("Sales", "mean")
    )
    .round(2)
)

cuadro_categoria['Margen_porcentaje'] = (
    (cuadro_categoria['Ganancia_Total'] / cuadro_categoria['Ventas_Totales']) * 100
).round(2)

cuadro_categoria = cuadro_categoria.sort_values("Margen_porcentaje",ascending= False)

print("\n--- CUADRO DE MANDO INTEGRAL POR CATEGORÍA ---")
print(cuadro_categoria)

analisis_descuentos = (
    ventas
    .groupby("Region")
    .agg(
        Ganancia_Promedio=("Profit", "mean"),
        Descuento_Promedio=("Discount", "mean")
    )
    .sort_values("Descuento_Promedio", ascending=False)
    .round(2)
)
print("\nAnálisis de Descuentos por Región")
print(analisis_descuentos)

matriz_region_categoria = (
    ventas
    .groupby(["Region","Category"])
    .agg(
        Ventas =("Sales","sum"),
        Ganancia =("Profit","sum")
    )
    .reset_index()
    .round(2)
)

print("\n--- ANÁLISIS CRUZADO: REGIÓN Y CATEGORÍA ---")
print(matriz_region_categoria)

subcat_perdidas = (
    ventas
    .groupby("Sub-Category")['Profit']
    .sum()
    .sort_values(ascending= True)
    .head(5)
)
print("\nTop 5 Sub-Categorías con Pérdidas:")
print(subcat_perdidas)

print(ventas.dtypes)
analisis_pareto = (
    ventas
    .groupby("Customer Name")
    .agg(
        Ventas_Totales=("Sales", "sum"),
        Total_Ordenes=("Order ID", "nunique")
    )
    .sort_values("Ventas_Totales", ascending=False)
)

# Cálculo de la suma global de ventas
venta_global = ventas["Sales"].sum()

# Porcentaje de aporte individual
analisis_pareto["Acumulado_porcentual"] = ((analisis_pareto["Ventas_Totales"] / venta_global) * 100).round(2)

print("\n--- ANÁLISIS DE PARETO DE CLIENTES (TOP 10) ---")
print(analisis_pareto.head(10))

clientes_top_10 = analisis_pareto[analisis_pareto["Acumulado_porcentual"] <= 10]

print("\nNúmero de clientes que generan el primer 10% de la venta total:")
print(len(clientes_top_10))



# print(f"\nNúmero de clientes que generan el primer 10% de la venta total: {len(clientes_top_10)}")

File: exercises/r/03_class.R
# library(tidyverse)
suppressPackageStartupMessages(library(tidyverse))

ventas <- read_csv("datasets/superstore.csv", locale = locale(encoding = "latin1"))

# print(names(ventas))
colnames(ventas) <- str_replace_all(colnames(ventas), " ", "_")
colnames(ventas) <- str_replace_all(colnames(ventas), "-", "_")
# print(names(ventas))

# Función
analizar_top_productos <- function(datos, variable_grupo, variable_orden, top_n = 10) {
  datos %>%
    group_by({{ variable_grupo }}) %>%
    summarise(
      Ventas = sum(Sales, na.rm = TRUE),
      Ganancia = sum(Profit, na.rm = TRUE),
      Cantidad = sum(Quantity, na.rm = TRUE)
    ) %>%
    arrange({{ variable_orden }}) %>%
    slice_head(n = top_n)
}

cat("\nVentas por categoría\n")

# ventas_categoria <- ventas %>%
#   group_by(Category) %>%
#   summarise(
#     Total_Sales = sum(Sales)
#   ) %>%
#   arrange(desc(Total_Sales))

# print(ventas_categoria)

conteo_ordenes <- ventas %>%
  count(Category) %>%
  arrange(desc(n))

  cat("\nConteo de transacciones por categoría:\n")
print(conteo_ordenes)

# Cuadro Categoría
cuadro_categoria <- ventas %>%
  group_by(Category) %>%
  summarise(
    Ventas_Totales   = sum(Sales, na.rm = TRUE),
    Ganancia_Total   = sum(Profit, na.rm = TRUE),
    Unidades_Vendidas = sum(Quantity, na.rm = TRUE),
    Ticket_Promedio  = mean(Sales, na.rm = TRUE)
  ) %>%
  mutate(
    Margen_porcentual = round((Ganancia_Total / Ventas_Totales) * 100, 2)
  ) %>%
  arrange(desc(Margen_porcentual))

cat("\n--- CUADRO DE MANDO INTEGRAL POR CATEGORÍA ---\n")
print(cuadro_categoria, width = Inf)

# Análisis de descuentos
analisis_descuentos <- ventas %>%
  group_by(Region) %>%
  summarise(
    Ganancia_Promedio  = mean(Profit, na.rm = TRUE),
    Descuento_Promedio = mean(Discount, na.rm = TRUE)
  ) %>%
  arrange(desc(Descuento_Promedio))

cat("\nAnálisis de Descuentos por Región:\n")
print(analisis_descuentos)

# Matriz Región y Categoría
matriz_region_categoria <- ventas %>%
  group_by(Region, Category) %>%
  summarise(
    Ventas   = sum(Sales, na.rm = TRUE),
    Ganancia = sum(Profit, na.rm = TRUE),
    .groups  = "drop"
  )  %>%
arrange(Ganancia)

cat("\n--- ANÁLISIS CRUZADO: REGIÓN Y CATEGORÍA ---\n")
print(matriz_region_categoria)

# Subcategorías con menor ganancia
subcat_perdidas <- ventas %>%
  group_by(Sub_Category) %>%
  summarise(Profit = sum(Profit, na.rm = TRUE)) %>%
  arrange(Profit) %>%
  slice_head(n = 5)

cat("\nTop 5 Sub-Categorías con Pérdidas:\n")
print(subcat_perdidas)

# Cuadro pareto
venta_global <- sum(ventas$Sales, na.rm = TRUE)

analisis_pareto <- ventas %>%
  group_by(Customer_Name) %>%
  summarise(
    Ventas_Totales = sum(Sales, na.rm = TRUE),
    Total_Ordenes  = n_distinct(Order_ID)
  ) %>%
  arrange(desc(Ventas_Totales)) %>%
  mutate(
    Aporte_porcentual    = round((Ventas_Totales / venta_global) * 100, 2),
    Acumulado_porcentual = round(cumsum(Aporte_porcentual), 2)
  )

cat("\n--- ANÁLISIS DE PARETO DE CLIENTES (TOP 10) ---\n")
print(head(analisis_pareto, 10))

# Clientes que representan el 10% de mis ventas
clientes_top_10 <- analisis_pareto %>%
  filter(Acumulado_porcentual <= 10)

cat("\nNúmero de clientes que generan el primer 10% de la venta total:"
, nrow(clientes_top_10), "\n")


# Analisis de productos
# analisis_productos <- ventas %>%
#  group_by(Product_Name) %>%
#  summarise(
#    Ventas = sum(Sales, na.rm = TRUE),
#    Ganancia = sum(Profit, na.rm = TRUE),
#    Cantidad = sum(Quantity, na.rm = TRUE)
#  ) %>%
#  arrange(Ganancia) %>%
#  slice_head(n = 10)

# cat("\nAnálisis de Productos:\n")
# print(analisis_productos)


cat("\nTop Productos\n")
print(analizar_top_productos(ventas, Product_Name, Ganancia, 5))

# Columna Fecha y introducir el tiempo  Order_Date
ventas <- ventas %>%
mutate(Order_Date = lubridate::mdy(Order_Date))

analisis_mensual <- ventas %>%
mutate(Mes = lubridate::month(Order_Date, label= TRUE)) %>%
group_by(Mes) %>%
summarise(
    Ventas = sum(Sales, na.rm = TRUE)
)

print(analisis_mensual)







